IPs/RSKIP336.md

```diff
@@ -0,0 +1,60 @@
+---
+rskip: 336
+title: Simple Parallelizable Semaphore
+description: 
+status: Draft
+purpose: Sca
+author: SDL (@sergiodemianlerner)
+layer: Core
+complexity: 2
+created: 2022-07-11
+---
+# Simple Parallelizable Semaphore
+
+|RSKIP          |336           |
+| :------------ |:-------------|
+|**Title**      |Simple Parallelizable Semaphore|
+|**Created**    |11-JUL-2022 |
+|**Author**     |SDL |
+|**Purpose**    |Sca |
+|**Layer**      |Core |
+|**Complexity** |2 |
+|**Status**     |Draft |
+|**discussions-to**     ||
+
+# **Abstract**
+
+This RSKIP proposes adding a precompiled contract that provides a semaphore function. This function aims to replace the typical semaphores implemented as contract  storage cells. The benefit is that the precompile-based semaphore does not write to storage, and therefore is is parallelizable under [RSKIP-144](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP144.md), while the typical sempaphore is not. 
+
+# **Motivation**
+
+The typical smart contract semaphore uses a contract storage cell. The algorithm is simple: the code checks that a storage cell is zero on entry, aborting if not, and then sets it to 1. After executing the required code, it resets the cell back to 0 before exiting. This is the algorithm implemented in OpenZeppelin's [ReentrancyGuard](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol). The algorithm results in a read-write pattern on the semaphore's storage cell. This pattern prevents the parallelization of the execution of the smart contract using [RSKIP-144](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP144.md).
+The only way to paralellize contracts using this semaphore contruction would be that the VM automatically detect that a variable is used as a sempahore, and proves that the sempahore behaves as required by static analysys. This is difficult for two reasons. First, the CPU cost of detection and/or proving may be high. Second, some contract functions may not be protected by the semaphore, meaning that some execution paths do not alter the semaphore, which may complicate proving. 
+Therefore this proposal aims to protect future contracts, rather than to protect already deployed ones.
+
+# **Specification**
+
+Starting from an activation block (TBD) a new precompiled contract `Semaphore` is created at address 0x01000011. when `Semaphore` is called, if the caller address is present more than once in the stack, the contract behaves as if the first instruction had been a REVERT, therefore the CALL returns 0. Otherwise, it executes no code and returns 1. The gas cost of the contract execution is set to 500, which is consumed independently of the call result.
```

Comment:
Maybe I'm missing something, but it isn't clear to me how it works and how it prevents two contracts writing at the same storage key

Reply:
If you mean "two parallel executing contracts writing to the same storage key" then this proposal does nothing to change that. That will be detected by the tracking of the storage key itself.
This RSKIP is not to prevent something, but to allow something. What is allowed is to make transaction dependency checks work only on storage keys, but not on the semaphore, which has a specific function that we know does not interfere with parallel processing.